### PR TITLE
Laboratory work #4, Maxim Matveev - 21FPL1 

### PR DESCRIPTION
**Pull request by yourMasya**
I don't know where I'm going
But I sure know where I've been
Hanging on the promises
In songs of yesterday

And I've made up my mind
I ain't wasting no more time

Here I go again
Here I go again

Though I keep searching for an answer
I never seem to find what I'm looking for
Oh, Lord, I pray
You give me strength to carry on

'Cause I know what it means
To walk along the lonely street of dreams

Here I go again on my own
Goin' down the only road I've ever known
Like a hobo I was born to walk alone

But I've made up my mind
I ain't wasting no more time

Just another heart in need of rescue
Waiting on love's sweet charity
I am gonna hold on
For the rest of my days

'Cause I know what it means
To walk along the lonely street of dreams

Here I go again on my own
Goin' down the only road I've ever known
Like a hobo I was born to walk alone
And I've made up my mind
I ain't wasting no more time

But here I go again
Here I go again
Here I go again
Here I go

And I've made up my mind
I ain't wasting no more time

Here I go again on my own
Goin' down the only road I've ever known
Like a hobo I was born to walk alone
'Cause I know what it means
To walk along the lonely street of dreams

Here I go again on my own
Goin' down the only road I've ever known
Like a hobo I was born to walk alone
I have made up my mind
I ain't wasting no more time

But here I go again
Here I go again
Here I go again
Here I go
Here I go again

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Looks like nothing to review(

**Comment by artyomtugaryov** [comment]
But the poem is great

File: lab_4_summarization_textrank/main.py [hunk 1]
```diff
@@ -2,16 +2,43 @@
...
 
 PreprocessedSentence = tuple[str, ...]
 EncodedSentence = tuple[int, ...]
 
 
+def check_types(collection: Any,
+                collections_type: list[Type]) \
+        -> None:
+    """
+    blegh
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:17]: improve

File: lab_4_summarization_textrank/main.py [hunk 2]
```diff
@@ -2,16 +2,43 @@
...
+                                                               and isinstance(collection, bool)))
+        checks.append(check)
+    if all(checks):
+        raise ValueError
+
+
+def check_items_type(collection: Any, collections_type: list[Type], items_type: list[Type]) \
+        -> None:
+    """
+    blegh
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:32]: .

File: lab_4_summarization_textrank/main.py [hunk 3]
```diff
@@ -20,59 +47,68 @@ def __init__(self, text: str, position: int) -> None:
         """
         Constructs all the necessary attributes
         """
-        pass
+
+        check_types(text, [str])
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:51]: check_types(text, str, tuple)

File: lab_4_summarization_textrank/main.py [hunk 4]
```diff
@@ -20,59 +47,68 @@ def __init__(self, text: str, position: int) -> None:
...
+        return self._preprocessed
 
     def set_encoded(self, encoded_sentence: EncodedSentence) -> None:
         """
         Sets the attribute
         :param encoded_sentence: the encoded sentence (a sequence of numbers)
         :return: None
         """
-        pass
+        check_items_type(encoded_sentence, [tuple], [int])
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:103]: .

File: lab_4_summarization_textrank/main.py [hunk 5]
```diff
@@ -120,23 +172,30 @@ def __init__(self) -> None:
...
     def _learn_indices(self, tokens: tuple[str, ...]) -> None:
         """
         Fills attributes mapping words and integer equivalents to each other
         :param tokens: a sequence of string tokens
         :return:
         """
-        pass
+        check_items_type(tokens, [tuple], [str])
+        new_tokens = (token for token in tokens if token not in self._word2id)
+        for ind, token in enumerate(new_tokens, start=max(1_000, 1_000 + len(self._word2id))):
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:185]: move to instance attributes

Comment (demid5111) [reply lab_4_summarization_textrank/main.py:185]: magic constant

File: lab_4_summarization_textrank/start.py [hunk 6]
```diff
@@ -2,6 +2,12 @@
 TextRank summarizer starter
 """
 from pathlib import Path
+from string import punctuation
+from lab_4_summarization_textrank.main import (
+    SentenceEncoder,
+    SentencePreprocessor, SimilarityMatrix,
+    TextRankSummarizer, Buddy
+) #NoRelevantTextsError
```
Comment (demid5111) [inline lab_4_summarization_textrank/start.py:10]: ??

File: lab_4_summarization_textrank/start.py [hunk 7]
```diff
@@ -27,6 +33,13 @@
...
 
-    RESULT = None
+    punctuation = tuple(punctuation)
+    PREPROCESSOR = SentencePreprocessor(stop_words, punctuation)
+    SENTENCES = PREPROCESSOR.get_sentences(text)
+    ENCODER = SentenceEncoder()
+    ENCODER.encode_sentences(SENTENCES)
+    print("Encoded:", *(sentence.get_encoded() for sentence in SENTENCES))
+
+    RESULT = True
```
Comment (demid5111) [inline lab_4_summarization_textrank/start.py:43]: [sentence.get_encoded() for sentence in SENTENCES]

**Review by demid5111 — CHANGES_REQUESTED** [review]

File: lab_4_summarization_textrank/main.py (hunk 1 above)
Comment (yourMasya) [reply lab_4_summarization_textrank/main.py]: solved)0))

File: lab_4_summarization_textrank/main.py (hunk 2 above)
Comment (yourMasya) [reply lab_4_summarization_textrank/main.py]: solved)0))

File: lab_4_summarization_textrank/main.py (hunk 4 above)
Comment (yourMasya) [reply lab_4_summarization_textrank/main.py]: solved)0)).

File: lab_4_summarization_textrank/main.py (hunk 5 above)
Comment (yourMasya) [reply lab_4_summarization_textrank/main.py]: solved)0))

File: lab_4_summarization_textrank/start.py (hunk 6 above)
Comment (yourMasya) [reply lab_4_summarization_textrank/start.py]: solved)0))

**Pull request closed without merging**